### PR TITLE
feat(zink-codegen): introduce derive macro for events 

### LINKED ISSUE #255: Improve the declaration of events
### Describe the feature

we currently only have a baseline event implementation, gather events as an enum, rich and dynamic `log{1,2,3,4}`

### Solution

```rust
#[derive(Event)]
enum Event {
  Transfer { from: Address, to: Address, value: U256 },
}
```

ABIs are also required

### PR DESCRIPTION
**Pull request by malik672**
Resolves #255 

- [x] generate ABIs for events and gathered events as enums
- [x] introduce util function for converting type to string

0x30bE4D758d86cfb1Ae74Ae698f2CF4BA7dC8d693

### CODE REVIEW

**Comment by malik672** [comment]
@clearloop I've been thinking how should I assert the logs

**Comment by malik672** [comment]
@clearloop

File: zink/codegen/src/event.rs [hunk 1]
```diff
@@ -1,19 +1,340 @@
...
+            Ok(format!("{}({})", variant_name, params))
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(quote! {
+        vec![
+            #(#variant_signatures.to_string()),*
+        ].join(";")
+    })
+}
```
Comment (clearloop) [inline zink/codegen/src/event.rs:291]: this has already been implemented here this has already been implemented in https://github.com/zink-lang/zink/blob/main/evm/abi/src/abi.rs#L28 just need to `Abi::parse` in this file

**Review by clearloop — CHANGES_REQUESTED** [review]
Sorry for the late! was busy these days, almost finished! your new tests are totally correct!

at the current stage, you now need to handle the log function in the visitor, see https://github.com/zink-lang/zink/blob/main/codegen/src/visitor/log.rs#L62

If you need more details or have specific questions, free feel to pin me in the telegram group! In case this issue is a bit harder than the previous, just updated the bounty to large level!

**Comment by malik672** [comment]
> Sorry for the late! was busy these days, almost finished! your new tests are totally correct!
> 
> at the current stage, you now need to handle the log function in the visitor, see https://github.com/zink-lang/zink/blob/main/codegen/src/visitor/log.rs#L62
> 
> If you need more details or have specific questions, free feel to pin me in the telegram group! In case this issue is a bit harder than the previous, just updated the bounty to large level!

yeah how will I do this ?

**Comment by malik672** [comment]
```    /// Log a message with topics.
    pub fn log(&mut self, count: usize) -> Result<()> {
        let topics = Vec::<Vec<u8>>::default();
 
        let (offset, size) = self.data()?;
        let size = size as usize;
        let data = self.env.data.load(offset, size)?;

        tracing::debug!("log1 name: {:?}", data);

        for topic in topics {
            self.masm.push(&topic)?;
        }

    
        // 1. write data to memory
        let MemoryInfo { offset, size } = self.masm.memory_write_bytes(&data)?;

        // 3. prepare the offset and size of the data.
        self.masm.push(&size.to_ls_bytes())?;
        self.masm.push(&offset)?;

        // 4. run log for the data
        match count {
            0 => self.masm._log0(),
            1 => self.masm._log1(),
            2 => self.masm._log2(),
            3 => self.masm._log3(),
            4 => self.masm._log4(),
            _ => unreachable!("invalid topics"),
        }?;

        Ok(())
    }``` @clearloop

**Comment by malik672** [comment]
> ```
>     pub fn log(&mut self, count: usize) -> Result<()> {
>         let topics = Vec::<Vec<u8>>::default();
>  
>         let (offset, size) = self.data()?;
>         let size = size as usize;
>         let data = self.env.data.load(offset, size)?;
> 
>         tracing::debug!("log1 name: {:?}", data);
> 
>         for topic in topics {
>             self.masm.push(&topic)?;
>         }
> 
>     
>         // 1. write data to memory
>         let MemoryInfo { offset, size } = self.masm.memory_write_bytes(&data)?;
> 
>         // 3. prepare the offset and size of the data.
>         self.masm.push(&size.to_ls_bytes())?;
>         self.masm.push(&offset)?;
> 
>         // 4. run log for the data
>         match count {
>             0 => self.masm._log0(),
>             1 => self.masm._log1(),
>             2 => self.masm._log2(),
>             3 => self.masm._log3(),
>             4 => self.masm._log4(),
>             _ => unreachable!("invalid topics"),
>         }?;
> 
>         Ok(())
>     }``` @clearloop 
> ```

**Comment by clearloop** [comment]
> ```rust
>     pub fn log(&mut self) -> Result<()> {
>         let data = self.env.data.load(offset, size)?;
>
>         // 1. write data to memory
>         let MemoryInfo { offset, size } = self.masm.memory_write_bytes(&data)?;
> 
>         // 3. prepare the offset and size of the data.
>         self.masm.push(&size.to_ls_bytes())?;
>         self.masm.push(&offset)?;
> 
>         // 4. run log for the data
>         match count {
>             0 => self.masm._log0(),
>             1 => self.masm._log1(),
>             2 => self.masm._log2(),
>             3 => self.masm._log3(),
>             4 => self.masm._log4(),
>             _ => unreachable!("invalid topics"),
>         }?;
> 
>         Ok(())
> ```

like this

**Comment by malik672** [comment]
> let (offset, size) = self.data()?;

quite similar to what i did first, made the tests fail

**Comment by malik672** [comment]
> > ```rust
> >     pub fn log(&mut self, count: usize) -> Result<()> {
> >         let data = self.env.data.load(offset, size)?;
> > 
> >         // 1. write data to memory
> >         let MemoryInfo { offset, size } = self.masm.memory_write_bytes(&data)?;
> > 
> >         // 3. prepare the offset and size of the data.
> >         self.masm.push(&size.to_ls_bytes())?;
> >         self.masm.push(&offset)?;
> > 
> >         // 4. run log for the data
> >         match count {
> >             0 => self.masm._log0(),
> >             1 => self.masm._log1(),
> >             2 => self.masm._log2(),
> >             3 => self.masm._log3(),
> >             4 => self.masm._log4(),
> >             _ => unreachable!("invalid topics"),
> >         }?;
> > 
> >         Ok(())
> > ```
> 
> like this

it sorts normally: --- STDERR:              zink::log log1 ---
thread 'log1' panicked at tests/log.rs:25:5:
assertion `left == right` failed
  left: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 112, 111, 110, 103]
 right: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, 105, 110, 103]

**Comment by clearloop** [comment]
> assertion `left == right` failed left: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 112, 111, 110, 103] right: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, 105, 110, 103]

looks closed! now you can decode this output to see what exactly the log is!

```
let left = String::from_lossy(&[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 112, 111, 110, 103]);
let right = String::from_lossy(&[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 80, 105, 110, 103])
```

**Comment by clearloop** [comment]
see my fix here: https://github.com/malik672/zink/compare/sync...zink-lang:zink:cl/sync

File: zink/codegen/Cargo.toml [hunk 2]
```diff
@@ -17,5 +17,6 @@ heck.workspace = true
 hex.workspace = true
 proc-macro2.workspace = true
 quote.workspace = true
+sha3 = "0.10.8"
```
Comment (clearloop) [inline zink/codegen/Cargo.toml:20]:
```suggestion

```

File: zink/codegen/src/event.rs [hunk 3]
```diff
@@ -1,19 +1,339 @@
...
+
+/// Validate field types
+fn validate_types(types: &[&Type]) -> Result<()> {
+    for ty in types {
+        if !is_supported_type(ty) {
+            return Err(Error::new_spanned(
+                ty,
+                format!("Unsupported type for event field: {}", quote!(#ty)),
+            ));
+        }
```
Comment (clearloop) [inline zink/codegen/src/event.rs:228]: matches!(sol_abi::Param::from(&ty.to_string()), )
```suggestion
       if matches!(sol_abi::Param::from(&ty.to_string()), sol_abi::Param::Unknown(_)) {
            return Err(Error::new_spanned(
                ty,
                format!("Unsupported type for event field: {}", quote!(#ty)),
            ));
       }
```

File: examples/log.rs [hunk 4]
```diff
@@ -1,99 +1,204 @@
...
-pub fn log1() {
-    Ping.log1(b"pong");
+pub mod event_tests {
+
+    use super::*;
+
+    /// Test log0
+    #[zink::external]
+    pub fn test_log0() {
+        unsafe { zink::ffi::evm::log0(b"MyEvent") }
```
Comment (clearloop) [inline examples/log.rs:29]:
```suggestion
        MyEvent::topic()
```

File: examples/log.rs [hunk 5]
```diff
@@ -1,99 +1,204 @@
...
+    }
+
+    /// Test log1
+    #[zink::external]
+    pub fn test_log1(value: U256) {
+        unsafe {
+            let topic = value.bytes32();
+            zink::ffi::evm::log1(topic, b"MyEvent")
```
Comment (clearloop) [inline examples/log.rs:37]:
```suggestion
            MyEvent::topic1(value)
```

File: examples/log.rs [hunk 6]
```diff
@@ -1,99 +1,204 @@
...
+        }
+    }
+
+    /// Test log2
+    #[zink::external]
+    pub fn test_log2(value1: U256, value2: U256) {
+        unsafe {
+            let topic1 = value1.bytes32();
+            let topic2 = value2.bytes32();
+            zink::ffi::evm::log2(topic1, topic2, b"MyEvent")
```
Comment (clearloop) [inline examples/log.rs:47]:
```suggestion
            MyEvent::topic2(topic1, topic2)
```

File: examples/log.rs [hunk 7]
```diff
@@ -1,99 +1,204 @@
...
+    }
+
+    /// Test log3
+    #[zink::external]
+    pub fn test_log3(value1: U256, value2: U256, value3: U256) {
+        unsafe {
+            let topic1 = value1.bytes32();
+            let topic2 = value2.bytes32();
+            let topic3 = value3.bytes32();
+            zink::ffi::evm::log3(topic1, topic2, topic3, b"MyEvent")
```
Comment (clearloop) [inline examples/log.rs:58]:
```suggestion
           MyLogic::topic3(value1.bytes32(), value2.bytes32(), value3.bytes32())
```

File: examples/log.rs [hunk 8]
```diff
@@ -1,99 +1,204 @@
...
+    /// Test log4
+    #[zink::external]
+    pub fn test_log4(value1: U256, value2: U256, value3: U256, value4: U256) {
+        unsafe {
+            let topic1 = value1.bytes32();
+            let topic2 = value2.bytes32();
+            let topic3 = value3.bytes32();
+            let topic4 = value4.bytes32();
+
+            zink::ffi::evm::log4(topic1, topic2, topic3, topic4, b"MyEvent")
```
Comment (clearloop) [inline examples/log.rs:71]: same as above

**Review by clearloop — CHANGES_REQUESTED** [review]
Everything works now!

Now we need to wrap these log functions into our proc-macro

for the event abi part, let's handle it in a separated PR (#299),  since it modifies logic in compiler

**Comment by clearloop** [comment]
need to generate sugar functions in our proc-macro, the implementation till [583dbe2](https://github.com/zink-lang/zink/pull/298/commits/583dbe204acf17c23ed745a98ee2a16783f29f43) makes sure our deep logic are working now, so the last thing is wrap them in proc-macro

**Comment by malik672** [comment]
@clearloop might need your help here

File: zink/codegen/src/event.rs [hunk 9]
```diff
@@ -1,19 +1,340 @@
...
-            const NAME: &'static [u8] = #name;
+        impl Event for #name {
+            const NAME: &'static [u8] = #name_bytes;
+
+
+            fn log0(&self) {
+                match self {
+                    #(#variant_implementations)*
+                }
+            }
```
Comment (clearloop) [inline zink/codegen/src/event.rs:74]: we don't need the `Event` trait anymore

```suggestion
        impl #name {
           fn [< #variant0:snake_case >]() {
               unsafe { zink::ffi::log() }
           }
           
           // ...
        }
```

**Review by clearloop — COMMENTED** [review]
Instead of `quote!`, construct the functions with the `ImplItemType` could make things more dynamic

File: zink/codegen/src/event.rs (hunk 9 above)
Comment (malik672) [reply zink/codegen/src/event.rs]: impl #name {
            pub fn log0(&self) {
                match self {
                    #(#variant_implementations)*
                }
            }

            pub const fn name() -> &'static [u8] {
                #name_bytes
            }
        }
        
        @clearloop something like this

Comment (malik672) [reply zink/codegen/src/event.rs]: @clearloop something like this :)

Comment (malik672) [reply zink/codegen/src/event.rs]: @clearloop something like this :)

Comment (clearloop) [reply zink/codegen/src/event.rs]: https://github.com/zink-lang/zink/pull/298/commits/a8a3eeca309de405e427ddfb1284ed2f62709e7d

**Review by clearloop — APPROVED** [review]
Thanks for contributing on this issue! however, I can not send you the full bounty  because I solved the issue instead finally XD, it's unfair to other contributors who solve tricky issues independently, hope you can understand!

Apart from that, thank you again for spending time on digging the feature! see the transaction 🎁 : https://polygonscan.com/tx/0x1b588cab3bdb014fe247acf4c06f825515af6b8ee4c1df1c505de220f7af39b1

File: examples/log.rs [hunk 10]
```diff
@@ -1,99 +1,133 @@
...
+pub enum MyEvent {
+    /// Event with one topic
+    Topic1(U256),
+    /// Event with two topics
+    Topic2(U256, U256),
+    /// Event with three topics
+    Topic3(U256, U256, U256),
+    /// Event with four topics
+    Topic4(U256, U256, U256, U256),
+}
```
Comment (CalabashSquash) [inline examples/log.rs:18]: How do you specify indexed topics?

Comment (g4titanx) [reply examples/log.rs:18]: @CalabashSquash  you can use the `#[indexed]` attribute on the fields you want to index within each variant. for example:
```rust
#[derive(Event)]
pub enum MyEvent {
    /// Event with one topic, indexed
    Topic1(#[indexed] U256),
    /// Event with two topics, first one indexed
    Topic2(#[indexed] U256, U256),
    /// Event with three topics, first two indexed
    Topic3(#[indexed] U256, #[indexed] U256, U256),
    /// Event with four topics, first three indexed
    Topic4(#[indexed] U256, #[indexed] U256, #[indexed] U256, U256),
}
```
the `#[indexed]` attribute marks a field as an indexed topic, which will be included in the event's topics for efficient filtering on the Ethereum blockchain. also, note that non-anonymous events can have up to 3 indexed fields, anonymous events can have up to 4 indexed fields and fields without `#[indexed]` are stored in the event's data payload, not as topics.
the logic lies [here](https://github.com/zink-lang/zink/blob/cd1bb5a355a0f37657484e52230ad022363f03ca/zink/codegen/src/event.rs#L75)

**Pull request merged**